### DISCUSSION TITLE
Is there a way to limit the size/nested of additionalProperties Map?

### DISCUSSION THREAD

**Discussion by KalCramer** [victools/jsonschema-generator#171] [Q&A] [answered]
Hello!

I'm adding `MAP_VALUES_AS_ADDITIONAL_PROPERTIES` option which allows the schema to accept `Map` values, but I was wondering if there is a way to limit the number of entries in the map like on a `List`. Is this something that is currently supported and I'm doing it incorrectly?

Something like the following maybe:
```java
@Size(max = 20)
private Map<String, Object> additionalParameters;
```

I would also want to make sure the values are not nested Maps. Is this something `patternProperties` would be able to do on top of this?

Edit: fixed wording

**Comment by CarstenWickner** [comment, marked as answer]
HI @KalCramer,

What the `Option.MAP_VALUES_AS_ADDITIONAL_PROPERTIES` does is this:
```java
configBuilder.forTypesInGeneral()
        .withAdditionalPropertiesResolver(scope -> {
            if (scope.getType().isInstanceOf(Map.class)) {
                // within a Map<Key, Value> allow additionalProperties of the Value type
                // if no type parameters are defined, this will result in additionalProperties to be omitted (by way of returning Object.class)
                return scope.getTypeParameterFor(Map.class, 1);
            }
            return null;
        });
```
You'd have to handle the scenario when `scope.getTypeParameterFor(Map.class, 1)` returns just `Object`, to define an alternative sub-schema that prevents nested Maps.
The `patternProperties` would only allow you to restrict the keys but not the values of the `Map`.

Regarding your second question:
The `@Size` annotation on anything but a `Collection` is ignored out-of-the-box.
In this context, it should affect the `maxProperties` attribute I imagine. For that you could add something like this:
```java
configBuilder.forFields()
        .withInstanceAttributeOverride((node, field, context) -> {
            Size annotation = field.getAnnotationConsideringFieldAndGetter(Size.class);
            if (annotation != null && x.max() < Integer.MAX_VALUE && field.getType().isInstanceOf(Map.class)) {
                node.put(context.getKeyword(SchemaKeyword.TAG_PROPERTIES_MAX), x.max());
            }
        });
```
